File: clients/zig/src/runtime.zig
//! What the generated half of this package reads and writes values through.
//!
//! Everything here is hand-written and never regenerated. It is the one seam between what the API
//! declares — the structs, the problems and the methods the generator writes under `generated/` —
//! and what it does not: how a JSON document is turned into a value, and what happens to a document
//! that does not say what it was declared to say.
//!
//! Reading is deliberately strict. A member the document declares as a string and the API answered
//! as a number stops the read rather than yielding a value whose documentation lies about what it
//! holds. Every failure of that kind is one of `DecodeError`, and the error is the message: Zig
//! errors carry no payload, so what a Ruby client would have said in a sentence is said by the name
//! of the error instead.
//!
//! Allocation is explicit and, more to the point, *owned*. Everything a decoded value points into —
//! the body that arrived, the document it was parsed into, every slice read out of it — is allocated
//! from one arena, and that arena travels with the value as an `Owned(T)`. One `deinit` frees the
//! lot; nothing is reachable afterwards, and nothing outlives it by accident. That is the shape the
//! ceilings of the shared corpus take here: a bound the other targets keep by discipline is a bound
//! the allocator applies, since nothing is read into memory the caller did not hand over.

const std = @import("std");

/// What the API answered is not what it declares it answers.
///
/// One error per way a document can fail to be what it says it is, rather than one error and a
/// string: the error is what a caller matches on, so it is what has to carry the meaning.
pub const DecodeError = error{
    /// The value read is not a JSON object, where a declared type was expected.
    NotAnObject,
    /// The value read is not a JSON array, where a list was expected.
    NotAnArray,
    /// The value read is not a string.
    NotAString,
    /// The value read is not a whole number, or does not fit the width it is declared at.
    NotAWholeNumber,
    /// The value read is not a number.
    NotANumber,
    /// The value read is not a boolean.
    NotABoolean,
    /// A member the document requires was not answered.
    MemberMissing,
    /// A string the document draws from a closed list is not one of the values it declares.
    ValueNotDeclared,
    /// The body is larger than this client reads.
    PayloadTooLarge,
    /// The body is not JSON.
    PayloadNotJson,
    /// There was no room to hold what was read.
    OutOfMemory,
};

/// What can go wrong writing a value back out, which is running out of room and nothing else.
pub const WriteError = std.mem.Allocator.Error;

/// Largest JSON document read out of a response body, in bytes. The transport caps what it reads off
/// a socket; this caps what is handed to the parser whichever way the bytes arrived.
pub const max_payload_bytes: usize = 8 * 1024 * 1024;

/// Deepest a JSON document may nest before the parser gives up, which is what keeps a document that
/// is nothing but brackets from growing the stack.
pub const max_payload_nesting: usize = 64;

/// Longest fragment of a response body a message carries. Bodies are answered by a server this
/// package does not control, so they are cut at a fixed budget rather than copied whole into
/// whatever the caller logs.
pub const max_preview_bytes: usize = 256;

/// An arena something is held in, which travels with what it holds rather than with the call.
///
/// The arena sits behind a pointer for the same reason [Owned]'s does: an allocator answered by an
/// arena holds the address of that arena, so an arena that moved would hand out an allocator
/// pointing at where it used to be.
///
/// Whoever asked for one frees it, once, with `deinit`. Nothing else refers to it.
pub const Kept = struct {
    arena: *std.heap.ArenaAllocator,

    pub fn init(allocator: std.mem.Allocator) std.mem.Allocator.Error!Kept {
        const arena = try allocator.create(std.heap.ArenaAllocator);
        arena.* = .init(allocator);
        return .{ .arena = arena };
    }

    /// Frees everything held in here, in one step.
    pub fn deinit(self: Kept) void {
        const child = self.arena.child_allocator;
        self.arena.deinit();
        child.destroy(self.arena);
    }
};

/// A value together with the arena everything it points into was allocated from.
///
/// The arena sits behind a pointer rather than inside this struct on purpose: an allocator answered
/// by an arena holds the address of that arena, so an arena that moved — which is exactly what
/// returning one by value does — would hand out an allocator pointing at where it used to be.
pub fn Owned(comptime T: type) type {
    return struct {
        arena: *std.heap.ArenaAllocator,
        value: T,

        const Self = @This();

        pub fn init(allocator: std.mem.Allocator) std.mem.Allocator.Error!Self {
            const arena = try allocator.create(std.heap.ArenaAllocator);
            arena.* = .init(allocator);
            return .{ .arena = arena, .value = undefined };
        }

        /// Frees the value and everything it points into, in one step.
        pub fn deinit(self: Self) void {
            const child = self.arena.child_allocator;
            self.arena.deinit();
            child.destroy(self.arena);
        }
    };
}

/// What one request the generated half issues is made of.
pub const Request = struct {
    method: []const u8,
    path: []const u8,
    query: []const QueryPair = &.{},
    body: ?std.json.Value = null,
};

/// What the API answered, which is a status and the bytes.
pub const Answer = struct {
    status: u16,
    payload: []const u8,
};

/// What one request is issued through.
///
/// A pointer and a function rather than an interface the generated half declares: nothing under
/// `generated/` knows what a socket is, and the transport that ships with this package is one
/// implementation of this among however many a caller writes.
pub const Transport = struct {
    context: *anyopaque,
    requestFn: *const fn (
        context: *anyopaque,
        allocator: std.mem.Allocator,
        request: Request,
    ) anyerror!Answer,

    pub fn request(
        self: Transport,
        allocator: std.mem.Allocator,
        asked: Request,
    ) anyerror!Answer {
        return self.requestFn(self.context, allocator, asked);
    }
};

/// One name and value travelling in a path or a query string.
pub const QueryPair = struct {
    name: []const u8,
    value: QueryValue,
};

/// A value as it travels in a request line, before anything is written.
///
/// `absent` is what an optional the caller passed nothing for reads as, which is what lets the
/// emitted method list every parameter the operation declares and leave the deciding to here.
pub const QueryValue = union(enum) {
    absent,
    text: []const u8,
    integer: i64,
    number: f64,
    boolean: bool,

    /// How this travels, written into `buffer` when it has to be written at all.
    pub fn written(self: QueryValue, buffer: []u8) ?[]const u8 {
        return switch (self) {
            .absent => null,
            .text => |carried| carried,
            .integer => |carried| std.fmt.bufPrint(buffer, "{d}", .{carried}) catch null,
            .number => |carried| std.fmt.bufPrint(buffer, "{d}", .{carried}) catch null,
            .boolean => |carried| if (carried) "true" else "false",
        };
    }
};

/// Longest a query value written out of a number may be.
pub const max_written_value_bytes: usize = 64;

/// That value, as one that travels in a request line.
///
/// Which of the shapes it takes is worked out from its type rather than said at the call site, so an
/// operation whose parameter is a number and one whose parameter is text are emitted the same way.
pub fn value(carried: anytype) QueryValue {
    const T = @TypeOf(carried);
    const info = @typeInfo(T);

    if (info == .optional) {
        return if (carried) |held| value(held) else .absent;
    }
    if (info == .null) {
        return .absent;
    }
    if (T == bool) {
        return .{ .boolean = carried };
    }
    if (info == .int or info == .comptime_int) {
        return .{ .integer = @intCast(carried) };
    }
    if (info == .float or info == .comptime_float) {
        return .{ .number = @floatCast(carried) };
    }
    return .{ .text = carried };
}

/// The characters a path segment carries as themselves; everything else travels percent-encoded.
fn unreserved(character: u8) bool {
    return switch (character) {
        'A'...'Z', 'a'...'z', '0'...'9', '-', '.', '_', '~' => true,
        else => false,
    };
}

/// That text, with nothing left in it that could name another segment or another parameter.
pub fn encoded(allocator: std.mem.Allocator, text_value: []const u8) WriteError![]const u8 {
    var out: std.ArrayList(u8) = .empty;
    errdefer out.deinit(allocator);

    for (text_value) |character| {
        if (unreserved(character)) {
            try out.append(allocator, character);
        } else {
            try out.print(allocator, "%{X:0>2}", .{character});
        }
    }
    return out.toOwnedSlice(allocator);
}

/// Where a request lands, with each placeholder of the template filled in.
pub fn path(
    allocator: std.mem.Allocator,
    template: []const u8,
    filled: []const QueryPair,
) WriteError![]const u8 {
    var filled_in: []const u8 = try allocator.dupe(u8, template);

    for (filled) |pair| {
        var buffer: [max_written_value_bytes]u8 = undefined;
        const carried = pair.value.written(&buffer) orelse "";
        const placeholder = try std.fmt.allocPrint(allocator, "{{{s}}}", .{pair.name});
        const segment = try encoded(allocator, carried);
        filled_in = try std.mem.replaceOwned(u8, allocator, filled_in, placeholder, segment);
    }

    return filled_in;
}

/// Whether that list of values carries the one asked about.
pub fn declares(values: []const []const u8, carried: []const u8) bool {
    for (values) |declared| {
        if (std.mem.eql(u8, declared, carried)) return true;
    }
    return false;
}

/// The type a reader answers, worked out from the reader itself.
///
/// What keeps every emitted decoder to one line: the type a member carries is already said by the
/// reader that reads it, so the call site does not repeat it.
pub fn ReaderValue(comptime reader: anytype) type {
    const info = @typeInfo(@TypeOf(reader)).@"fn";
    return @typeInfo(info.return_type.?).error_union.payload;
}

/// The members of an object the document declares, under the name it declares it with.
///
/// The name is not read at run time — a Zig error carries nothing for it to travel in — and it is
/// asked for anyway, so that a reader of an emitted decoder can see which type it decodes without
/// looking anywhere else.
pub fn asFields(carried: std.json.Value, owner: []const u8) DecodeError!std.json.ObjectMap {
    _ = owner;
    return switch (carried) {
        .object => |fields| fields,
        else => error.NotAnObject,
    };
}

/// A member the document requires, which is therefore missing when it is absent.
pub fn read(
    allocator: std.mem.Allocator,
    fields: std.json.ObjectMap,
    key: []const u8,
    comptime reader: anytype,
) DecodeError!ReaderValue(reader) {
    const carried = fields.get(key) orelse return error.MemberMissing;
    return reader(allocator, carried);
}

/// A member the document does not require, absent as readily as answered as null.
pub fn maybe(
    allocator: std.mem.Allocator,
    fields: std.json.ObjectMap,
    key: []const u8,
    comptime reader: anytype,
) DecodeError!?ReaderValue(reader) {
    const carried = fields.get(key) orelse return null;
    if (carried == .null) return null;
    return try reader(allocator, carried);
}

/// A string, refusing what merely spells like one.
pub fn text(allocator: std.mem.Allocator, carried: std.json.Value) DecodeError![]const u8 {
    _ = allocator;
    return switch (carried) {
        .string, .number_string => |held| held,
        else => error.NotAString,
    };
}

/// A whole number that fits in thirty-two bits.
pub fn integer32(allocator: std.mem.Allocator, carried: std.json.Value) DecodeError!i32 {
    const held = try integer64(allocator, carried);
    return std.math.cast(i32, held) orelse error.NotAWholeNumber;
}

/// A whole number. A boolean is not one, here or on the wire, and neither is a number the document
/// wrote with a fractional part.
pub fn integer64(allocator: std.mem.Allocator, carried: std.json.Value) DecodeError!i64 {
    _ = allocator;
    return switch (carried) {
        .integer => |held| held,
        else => error.NotAWholeNumber,
    };
}

/// A number, whether the document wrote it with a fractional part or not.
pub fn number(allocator: std.mem.Allocator, carried: std.json.Value) DecodeError!f64 {
    _ = allocator;
    return switch (carried) {
        .float => |held| held,
        .integer => |held| @floatFromInt(held),
        else => error.NotANumber,
    };
}

/// A boolean, refusing the numbers that stand in for one elsewhere.
pub fn boolean(allocator: std.mem.Allocator, carried: std.json.Value) DecodeError!bool {
    _ = allocator;
    return switch (carried) {
        .bool => |held| held,
        else => error.NotABoolean,
    };
}

/// A value the document does not describe, which is therefore kept as it arrived.
pub fn jsonValue(allocator: std.mem.Allocator, carried: std.json.Value) DecodeError!std.json.Value {
    _ = allocator;
    return carried;
}

/// Every item of an array, each one read the same way.
pub fn list(comptime reader: anytype) type {
    return struct {
        pub fn read(
            allocator: std.mem.Allocator,
            carried: std.json.Value,
        ) DecodeError![]const ReaderValue(reader) {
            const items = switch (carried) {
                .array => |held| held,
                else => return error.NotAnArray,
            };

            const out = try allocator.alloc(ReaderValue(reader), items.items.len);
            for (items.items, 0..) |item, index| {
                out[index] = try reader(allocator, item);
            }
            return out;
        }
    };
}

/// An object whose keys the document leaves open, as the pairs it carries.
///
/// A slice of pairs rather than a hash map: what arrives is read once and looked at, the order it
/// arrived in is the order it goes back out in, and nothing has to be hashed to write it.
pub fn Map(comptime T: type) type {
    return struct {
        entries: []const Entry = &.{},

        pub const Entry = struct { key: []const u8, value: T };
        const Self = @This();

        /// What that key carries, when the document carried it.
        pub fn get(self: Self, key: []const u8) ?T {
            for (self.entries) |entry| {
                if (std.mem.eql(u8, entry.key, key)) return entry.value;
            }
            return null;
        }

        /// Write one back the way the API reads it.
        pub fn toJson(self: Self, allocator: std.mem.Allocator) WriteError!std.json.Value {
            var out: std.json.ObjectMap = .empty;
            for (self.entries) |entry| {
                try out.put(allocator, entry.key, try written(allocator, entry.value));
            }
            return .{ .object = out };
        }
    };
}

/// Every value of an object whose keys the document leaves open.
pub fn map(comptime reader: anytype) type {
    return struct {
        pub fn read(
            allocator: std.mem.Allocator,
            carried: std.json.Value,
        ) DecodeError!Map(ReaderValue(reader)) {
            const fields = switch (carried) {
                .object => |held| held,
                else => return error.NotAnObject,
            };

            const Held = Map(ReaderValue(reader));
            const out = try allocator.alloc(Held.Entry, fields.count());
            var walked = fields.iterator();
            var index: usize = 0;
            while (walked.next()) |entry| : (index += 1) {
                out[index] = .{
                    .key = entry.key_ptr.*,
                    .value = try reader(allocator, entry.value_ptr.*),
                };
            }
            return .{ .entries = out };
        }
    };
}

/// One of the values a closed list declares, refusing anything the list does not carry.
pub fn memberOf(comptime declared: type) type {
    return struct {
        pub fn read(
            allocator: std.mem.Allocator,
            carried: std.json.Value,
        ) DecodeError![]const u8 {
            const held = try text(allocator, carried);
            if (!declared.member(held)) return error.ValueNotDeclared;
            return held;
        }
    };
}

/// That value, as the JSON it goes out as.
///
/// Which shape it takes is worked out from its type rather than said at the call site, which is what
/// keeps an emitted writer to one line per member whatever the member is.
pub fn written(allocator: std.mem.Allocator, carried: anytype) WriteError!std.json.Value {
    const T = @TypeOf(carried);
    const info = @typeInfo(T);

    if (T == std.json.Value) return carried;
    if (info == .optional) {
        return if (carried) |held| written(allocator, held) else .null;
    }
    if (info == .null) return .null;
    if (T == bool) return .{ .bool = carried };
    if (info == .int or info == .comptime_int) return .{ .integer = @intCast(carried) };
    if (info == .float or info == .comptime_float) return .{ .float = @floatCast(carried) };
    if (comptime isText(T)) return .{ .string = carried };

    if (info == .@"struct" and @hasDecl(T, "toJson")) {
        return carried.toJson(allocator);
    }
    if (info == .pointer and info.pointer.size == .slice) {
        var items: std.json.Array = .init(allocator);
        try items.ensureTotalCapacity(carried.len);
        for (carried) |item| {
            items.appendAssumeCapacity(try written(allocator, item));
        }
        return .{ .array = items };
    }

    @compileError("nothing says how to write a " ++ @typeName(T) ++ " back the way the API reads it");
}

/// Whether that type is the one this package carries text in.
fn isText(comptime T: type) bool {
    const info = @typeInfo(T);
    if (info != .pointer or info.pointer.size != .slice) return false;
    return info.pointer.child == u8;
}

/// Writes one member of a document, and writes nothing at all when it carries nothing.
///
/// A member the document does not require is absent rather than null when the caller passed nothing:
/// what goes out is what the API reads, and the two are not the same thing.
pub fn put(
    out: *std.json.ObjectMap,
    allocator: std.mem.Allocator,
    key: []const u8,
    carried: anytype,
) WriteError!void {
    if (@typeInfo(@TypeOf(carried)) == .optional) {
        const held = carried orelse return;
        try out.put(allocator, key, try written(allocator, held));
        return;
    }
    try out.put(allocator, key, try written(allocator, carried));
}

/// The JSON document a response body carries.
pub fn decodePayload(
    allocator: std.mem.Allocator,
    payload: []const u8,
) DecodeError!std.json.Value {
    if (payload.len > max_payload_bytes) return error.PayloadTooLarge;

    return std.json.parseFromSliceLeaky(std.json.Value, allocator, payload, .{
        .max_value_len = max_payload_bytes,
    }) catch |failed| switch (failed) {
        error.OutOfMemory => error.OutOfMemory,
        else => error.PayloadNotJson,
    };
}

/// The problem document a body names, or nothing when this client cannot read one out of it.
///
/// A body that is not the problem shape is a body naming no problem, rather than a second failure on
/// top of the first.
pub fn problemOf(
    comptime T: type,
    allocator: std.mem.Allocator,
    payload: []const u8,
) ?T {
    const document = decodePayload(allocator, payload) catch return null;
    return T.fromJson(allocator, document) catch null;
}

/// As much of a response body as a message may carry.
pub fn preview(payload: []const u8) []const u8 {
    if (payload.len <= max_preview_bytes) return payload;
    return payload[0..max_preview_bytes];
}

/// What to say about an answer the API document does not describe.
pub fn unreadable(
    allocator: std.mem.Allocator,
    status: u16,
    payload: []const u8,
) WriteError![]const u8 {
    return std.fmt.allocPrint(
        allocator,
        "the API answered {d} with a body this client cannot read: {s}",
        .{ status, preview(payload) },
    );
}

/// What to say about a problem the API reported.
pub fn reported(
    allocator: std.mem.Allocator,
    status: u16,
    payload: []const u8,
) WriteError![]const u8 {
    return std.fmt.allocPrint(
        allocator,
        "the API answered {d}: {s}",
        .{ status, preview(payload) },
    );
}

File: clients/lua/src/runtime.lua
--- What the generated half of this rock reads and writes values through.
---
--- Everything here is hand-written and never regenerated. It is the one seam between what the API
--- declares — the tables, the problems and the methods the generator writes under `generated/` — and
--- what it does not: how a JSON document is turned into a value, and what happens to a document that
--- does not say what it was declared to say.
---
--- Reading is deliberately strict. A member the document declares as a string and the API answered
--- as a number stops the read with the name of that member, rather than yielding a value whose
--- documentation lies about what it holds. Every failure of that kind is a `DecodeError`, so a
--- caller has one kind to match whatever the shape of the answer was.
---
--- A reader is a function of one value. The scalar ones are fields, since there is exactly one of
--- each; the ones built around another reader are functions, since there is one per shape.

local Errors = require("hook0.errors")
local Json = require("hook0.json")

local Runtime = {}

--- What the API answered is not what it declares it answers.
Runtime.DecodeError = Errors.kind("DecodeError", Errors.ClientError)

--- Longest fragment of a response body a message carries. Bodies are answered by a server this rock
--- does not control, so they are cut at a fixed budget rather than echoed whole into whatever the
--- caller logs.
Runtime.MAX_PREVIEW_BYTES = 256

--- Largest JSON document read out of a response body, in bytes. The transport caps what it reads off
--- a socket; this caps what is handed to the parser whichever way the bytes arrived.
Runtime.MAX_PAYLOAD_BYTES = 8 * 1024 * 1024

--- Deepest a JSON document may nest before the parser gives up.
Runtime.MAX_PAYLOAD_NESTING = Json.MAX_DEPTH

--- Most values one query string carries, which bounds what one request line costs.
Runtime.MAX_QUERY_PAIRS = 128

local function refuse(message)
  Errors.throw(Runtime.DecodeError, message)
end

--- What a value is, in the words a message uses.
local function named_type(value)
  if value == Json.null then
    return "null"
  end
  return type(value)
end

--- A string, refusing what merely spells like one.
--- @param value any
--- @return string
function Runtime.TEXT(value)
  if type(value) ~= "string" then
    refuse("expected a string, got " .. named_type(value))
  end
  return value
end

--- A whole number. A boolean is not one, here or on the wire, and neither is a number the document
--- wrote with a fractional part.
--- @param value any
--- @return integer
function Runtime.INTEGER(value)
  if math.type(value) ~= "integer" then
    refuse("expected a whole number, got " .. named_type(value))
  end
  return value
end

--- A number, whether the document wrote it with a fractional part or not.
--- @param value any
--- @return number
function Runtime.NUMBER(value)
  if type(value) ~= "number" then
    refuse("expected a number, got " .. named_type(value))
  end
  return value + 0.0
end

--- A boolean, refusing the numbers that stand in for one elsewhere.
--- @param value any
--- @return boolean
function Runtime.BOOLEAN(value)
  if type(value) ~= "boolean" then
    refuse("expected a boolean, got " .. named_type(value))
  end
  return value
end

--- A value the document does not describe, which is therefore kept as it arrived.
--- @param value any
--- @return any
function Runtime.JSON_VALUE(value)
  return value
end

--- As much of a response body as a message may carry.
--- @param payload any
--- @return string
function Runtime.preview(payload)
  local written = tostring(payload)
  if #written <= Runtime.MAX_PREVIEW_BYTES then
    return written
  end
  return written:sub(1, Runtime.MAX_PREVIEW_BYTES) .. "…"
end

--- What to say about an answer the API document does not describe.
--- @param status integer
--- @param payload string
--- @return string
function Runtime.unreadable(status, payload)
  return "the API answered " .. tostring(status) .. " with a body this client cannot read: " .. Runtime.preview(payload)
end

--- What to say about a problem the API reported.
--- @param status integer
--- @param problem table the problem document the API answered
--- @return string
function Runtime.reported(status, problem)
  local ok, written = Json.try_encode(problem:to_table())
  return "the API answered " .. tostring(status) .. ": " .. (ok and Runtime.preview(written) or "a problem it carried")
end

--- The JSON document a response body carries.
--- @param payload string
--- @return any
function Runtime.decode_payload(payload)
  if type(payload) ~= "string" then
    refuse("the response is " .. type(payload) .. ", not a body")
  end
  if #payload > Runtime.MAX_PAYLOAD_BYTES then
    refuse("the response is " .. #payload .. " bytes, above the " .. Runtime.MAX_PAYLOAD_BYTES .. " accepted")
  end

  local ok, read = Json.try_decode(payload)
  if not ok then
    local reason = Json.reason(read) or Errors.message(read)
    refuse("the response is not JSON: " .. Runtime.preview(payload) .. " (" .. reason .. ")")
  end
  return read
end

--- The problem document a body names, or nothing when this client cannot read one out of it.
---
--- What the generated half calls before it decides which failure to raise: a body that is not the
--- problem shape is a body naming no problem, rather than a second failure on top of the first.
---
--- @param schema table the generated table for the error contract
--- @param payload string the body the API answered
--- @return table|nil
function Runtime.problem_of(schema, payload)
  local ok, read = pcall(function()
    return schema.from_json(Runtime.decode_payload(payload))
  end)
  if ok then
    return read
  end
  return nil
end

--- The members of an object the document declares, under the name it declares it with.
--- @param value any
--- @param owner string what the document calls the object being read
--- @return table
function Runtime.as_fields(value, owner)
  if type(value) ~= "table" or Json.is_array(value) or value == Json.null then
    refuse(owner .. " is not a JSON object")
  end
  return value
end

--- Reads a member, saying which member it was that could not be read.
local function named(key, read)
  local ok, answered = pcall(read)
  if ok then
    return answered
  end
  if Errors.is(answered, Runtime.DecodeError) then
    refuse("`" .. key .. "`: " .. answered.message)
  end
  error(answered, 0)
end

--- A member the document requires, which is therefore missing when it is absent.
--- @param fields table
--- @param key string the name the member travels under
--- @param reader function
--- @return any
function Runtime.read(fields, key, reader)
  local carried = fields[key]
  if carried == nil then
    refuse("`" .. key .. "` is required and was not answered")
  end
  return named(key, function()
    return reader(carried)
  end)
end

--- A member the document does not require, absent as readily as answered as null.
--- @param fields table
--- @param key string the name the member travels under
--- @param reader function
--- @return any|nil
function Runtime.maybe(fields, key, reader)
  local carried = fields[key]
  if carried == nil or carried == Json.null then
    return nil
  end
  return named(key, function()
    return reader(carried)
  end)
end

--- Every item of an array, each one read the same way.
--- @param reader function
--- @return function
function Runtime.list(reader)
  return function(value)
    if type(value) ~= "table" or Json.is_object(value) or value == Json.null then
      refuse("expected an array, got " .. named_type(value))
    end
    local read = {}
    for index = 1, #value do
      read[index] = reader(value[index])
    end
    return read
  end
end

--- Every value of an object whose keys the document leaves open.
--- @param reader function
--- @return function
function Runtime.map(reader)
  return function(value)
    if type(value) ~= "table" or Json.is_array(value) or value == Json.null then
      refuse("expected an object, got " .. named_type(value))
    end
    local read = {}
    for key, carried in pairs(value) do
      read[Runtime.TEXT(key)] = reader(carried)
    end
    return read
  end
end

--- One of the values a closed list declares, refusing anything the list does not carry.
--- @param declared table the table the generator wrote for that list
--- @return function
function Runtime.member_of(declared)
  return function(value)
    local text = Runtime.TEXT(value)
    if not declared.member(text) then
      refuse("`" .. text .. "` is not one of the values that list declares")
    end
    return text
  end
end

--- Whether that list of values carries the one asked about.
--- @param values table
--- @param value any
--- @return boolean
function Runtime.declares(values, value)
  for index = 1, #values do
    if values[index] == value then
      return true
    end
  end
  return false
end

--- That table, as the JSON object it is meant to go out as.
--- @param value table
--- @return table
function Runtime.document(value)
  return Json.object(value)
end

--- A value that travels as it stands.
--- @param value any
--- @return any
function Runtime.itself(value)
  return value
end

--- An emitted value, written back the way the API reads it.
--- @param value table|nil
--- @return table|nil
function Runtime.written(value)
  if value == nil then
    return nil
  end
  return value:to_table()
end

--- Every item of a list, written back and marked as the array it is.
---
--- A table a caller assembled by hand says nothing about whether it is a list or a map, so it is
--- rebuilt rather than passed on: what goes out is what the API reads, whichever way it was built.
---
--- @param value table|nil
--- @param write function
--- @return table|nil
function Runtime.written_list(value, write)
  if value == nil then
    return nil
  end
  local out = {}
  for index = 1, #value do
    out[index] = write(value[index])
  end
  return Json.array(out)
end

--- Every value of a map, written back and marked as the object it is.
--- @param value table|nil
--- @param write function
--- @return table|nil
function Runtime.written_map(value, write)
  if value == nil then
    return nil
  end
  local out = {}
  for key, carried in pairs(value) do
    out[key] = write(carried)
  end
  return Json.object(out)
end

--- Whether two documents carry the same members, however deeply they nest.
--- @param left any
--- @param right any
--- @param depth integer|nil
--- @return boolean
function Runtime.same(left, right, depth)
  local at = depth or 1
  if at > Runtime.MAX_PAYLOAD_NESTING then
    return false
  end
  if left == right then
    return true
  end
  if type(left) ~= "table" or type(right) ~= "table" then
    return false
  end

  for key, carried in pairs(left) do
    if not Runtime.same(carried, right[key], at + 1) then
      return false
    end
  end
  for key in pairs(right) do
    if left[key] == nil then
      return false
    end
  end
  return true
end

--- Whether two emitted values carry the same members, which is what `==` answers on one.
--- @param left any
--- @param right any
--- @return boolean
function Runtime.equality(left, right)
  local function documented(value)
    if type(value) == "table" and type(value.to_table) == "function" then
      return value:to_table()
    end
    return value
  end
  return Runtime.same(documented(left), documented(right))
end

--- The characters a path segment carries as themselves; everything else travels percent-encoded.
local UNRESERVED = "[^A-Za-z0-9%-%._~]"

--- How a value travels in a request line, which is not always how Lua prints it.
--- @param value any
--- @return string
function Runtime.text_of(value)
  if type(value) == "boolean" then
    return tostring(value)
  end
  if math.type(value) == "float" and value == math.floor(value) and math.abs(value) < 2 ^ 53 then
    return string.format("%d", value)
  end
  return tostring(value)
end

--- A value as one segment of a path, with nothing left in it that could name another one.
--- @param value any
--- @return string
function Runtime.path_segment(value)
  return (Runtime.text_of(value):gsub(UNRESERVED, function(character)
    return string.format("%%%02X", character:byte())
  end))
end

--- Where a request lands, with each placeholder of the template filled in.
--- @param template string the path as the document writes it, placeholders included
--- @param filled table the value each placeholder carries
--- @return string
function Runtime.path(template, filled)
  local written = template
  for name, value in pairs(filled or {}) do
    if value == nil then
      refuse("`" .. name .. "` fills a segment of the path and carries nothing")
    end
    written = written:gsub("{" .. name:gsub("%W", "%%%0") .. "}", (Runtime.path_segment(value):gsub("%%", "%%%%")))
  end
  return written
end

--- What travels in the query string: every pair the operation declares that carries a value.
---
--- The pairs are what the emitted method lists, and this is what leaves out the ones the caller
--- passed nothing for — so an optional parameter costs nothing at the call site and nothing on the
--- wire.
---
--- @param pairs_of table|nil name and value pairs, in the order the operation declares them
--- @return table
function Runtime.query(pairs_of)
  local asked = {}
  if pairs_of == nil then
    return asked
  end
  if #pairs_of > Runtime.MAX_QUERY_PAIRS then
    refuse("a request carries more than the " .. Runtime.MAX_QUERY_PAIRS .. " query values accepted")
  end

  for index = 1, #pairs_of do
    local name, value = pairs_of[index][1], pairs_of[index][2]
    if value ~= nil then
      asked[#asked + 1] = { name, Runtime.text_of(value) }
    end
  end
  return asked
end

return Runtime
